### PR TITLE
Fix #1586: Randomize the initial version of an optimized MethodImpl.

### LINKED ISSUE #1586: Changing abstract class to trait causes "ReferenceError: XXX is not defined"
Hello,

So, I have seen this bug and spent a while to get a reproducible scenario. I would advice to follow exactly the procedure below as it seems even a small variation causes the bug not to be triggered.

* So to begins with, clone the repo of the [scala-js-example](https://github.com/sjrd/scala-js-example-app) (I'm not sure how previously cached data alter the behavior, so I would advice to really take a freshly cloned version) (by the way, I'm on the current master, aka commit e80339df4d25272d1dd4595ffbfadf188ecc4a90).

* In a terminal, cd into the dir and launch ```sbt ~fastOptJS```.

* Open the file ```src/main/scala/example/ScalaJSExample.scala``` with your text editor, at the end of the main function (inside the function, not after), add the following lines:
```scala
abstract class A(c: Int)
class B(d: Int) extends A(1)

val data = new B(2)
```

* Save the file (which should trigger the recompilation in the SBT console, wait for it to complete)

* Open the file ```index-fastopt.html``` with your browser (the bug is visible with Firefox and Chrome). Open the console of the web tools of your browser, everything should be Ok (no error in the console).

* Go back in the text editor, change the previous lines to:
```scala
trait A
class B(d: Int) extends A

val data = new B(2)
```

* Save the file (again, wait for the recompilation)

* Refresh the page in your browser and look at the console again, containing an error (the page should still display "It works!" though).
On Firefox: ```ReferenceError: $c_Lexample_ScalaJSExample$A$1 is not defined```
On Chrome: ```Uncaught ReferenceError: $c_Lexample_ScalaJSExample$A$1 is not defined```

I'm not too sure about the actual issue, but from what I have seen in the JavaScript code, it seems that the constructor of B is still trying to call the constructor of A (which does not exist anymore at this point), causing the error.

[edit]: The code throwing the error (second line):
```javascript
$c_Lexample_ScalaJSExample$B$1.prototype.init___I = (function(d) {
  $c_Lexample_ScalaJSExample$A$1.prototype.init___I.call(this, 1);
  return this
});
```

### PR DESCRIPTION
**Pull request by sjrd**
If we do not randomize the version number upon creation, we can
delete a MethodImpl from one place in the tree and recreate it in
a different place, with exactly the same name (if the parent
hiearchy of its enclosing class changes), which causes the Emitter
to think the method has not changed, whereas in fact it has nothing
to do with the previously deleted method.

### CODE REVIEW

**Comment by sjrd** [comment]
Review by @gzm0 
Tested locally, since we don't have tests for the incremental behavior of the linking pipeline.

**Comment by gzm0** [comment]
I need to think of this. Isn't this a more general issue in how we define versioning in the pipeline?

**Comment by sjrd** [comment]
Maybe, yes. In theory, everything should be about hashes, and have git-like versioning. In practice, we don't want to spend too much time hashing and re-hashing stuff. Drawing the line at the right place is hard.

**Comment by gzm0** [comment]
We need not randomize method versions to make this work. IIUC the problem is, that the incremental optimizer considers B extends A and B to be two different classes, whereas the rest of the compiler pipeline considers them to be the same classes.

If we make method version numbers increasing over class name and not optimizer class instance, this should be solved.

I hope this is somewhat clear...

**Comment by gzm0** [comment]
> everything should be about hashes, and have git-like versioning

I don't agree here. It is enough if each phase ensures that the version it emits changes, if the class/method it emits changes.

Since this is not an if and only if relation, this is relatively easy to solve since each phase can just bump the version of everything that the preceding phase changed (and the dependencies). If the incoming differences do not lead to no outgoing difference, it is up to the phase to decide whether it is faster to detect that, or just safely approximate and flag a change anyway.

File: tools/shared/src/main/scala/org/scalajs/core/tools/optimizer/GenIncOptimizer.scala [hunk 1]
```diff
@@ -696,7 +705,7 @@ abstract class GenIncOptimizer(semantics: Semantics,
     private[this] var _deleted: Boolean = false
 
     var lastInVersion: Option[String] = None
-    var lastOutVersion: Int = 0
+    var lastOutVersion: Int = randomByThread.get.nextInt()
```
Comment (gzm0) [inline tools/shared/src/main/scala/org/scalajs/core/tools/optimizer/GenIncOptimizer.scala]: Basically, here we need something like:

```scala
var lastOutVersion: Int = outVersions.getOrElse(owner.encodedName, 0).getOrElse(encodedName, 0)
```

Further, something that stores the outVersion when we bump it.

**Comment by sjrd** [comment]
Hum ... The problem is deeper than I thought.
You are right: it is sufficient for the optimizer to assign versions that differ if (but not *only if*) the content changes. *And that is what it did!*
The problem is that the optimizer and the emitter do not agree *on the very identity* of a method. For the optimizer, the identity of a method is its name + the entire entire list of class names in the parent hierarchy of its class. As such, if a method "moves" in the hierarchy, it effectively becomes a *different* method. Or should I say, it is deleted, and a new, unrelated method is created. And then this new method begins its versioning at 0, which is correct since it is a brand new method.
The the emitter, on the other hand, thinks the identity of a method is only its name + the name of its enclosing class. But *not* the classes in the hierarchy. It then thinks the new method is the same as the old method, and since they have both version 0, it has not changed. Which, from its point of view, is correct.

Hence, both phases are actually correct wrt to their definition of the identity of a method. But since they don't agree on that very definition, things break. The proper fix should be to make them agree on that, probably make the emitter think like the optimizer (not the other way around).

**Comment by sjrd** [comment]
Superseded by #1617.

**Pull request closed without merging**